src/Revise.jl

```diff
@@ -94,6 +94,74 @@ This list gets populated by callbacks that watch directories for updates.
 """
 const revision_queue = Set{Tuple{PkgData,String}}()
 
+const revision_event = Condition()
+
+"""
+    Revise.user_callbacks_queue
+
+Global variable, `user_callbacks_queue` holds `key` values for which the
+file has changed but the user hooks have not yet been called.
+"""
+const user_callbacks_queue = Set{Any}()
+
+const user_callbacks_by_file = Dict{String, Set{Any}}()
+const user_callbacks_by_key = Dict{Any, Any}()
+
+function add_callback(f, files, modules=nothing; key=gensym())
+    remove_callback(key)
```

Comment:
This is one valid option; another would be to throw if `key` has been previously used. I don't have strong feelings, but what was your thinking here?

Reply:
My thinking was to make the operation idempotent. I'm thinking this will be used in situations immediately after loading some data / code from a file in `files`. The operation of loading that data is presumably idempotent so it would be nice if adding `add_callback` right after the idempotent operation is idempotent as well. What do you think?

Reply:
To make it more explicit, and to improve the example snippet from before, here's how I'd use it:
```julia
const COUPLING_CONSTANTS = Ref(read_json("coupling_constants.json"))
Revise.add_callback(["coupling_constants.json"], key=:COUPLING_CONSTANTS) do
    COUPLING_CONSTANTS[] = read_json("coupling_constants.json")
end
```
so it doesn't matter how often this code executes (in particular, maybe it executes in response to one of these callbacks!). It would be even better to have the key depend on `@__MODULE__` in some way, or whatever else makes it a true global primary key in the user's situation.